### PR TITLE
Check type annotations in CI

### PR DESCRIPTION
**Pull request by codeinthehole**
This PR adds type annotations and checks they are satisfied in CI.

### CODE REVIEW

File: tests/test_ranges.py [hunk 1]
```diff
@@ -289,7 +292,7 @@ def test_range_difference_and_intersection_form_partition(a: ranges.Range, b: ra
             assert a_difference & b_difference is None
             assert a_difference.is_disjoint(b_difference)
             assert b_difference.is_disjoint(a_difference)
-            assert (a_difference | intersection | b_difference) == (a | b)
+            assert (a_difference | intersection | b_difference) == (a | b)  # type: ignore[operator]
```
Comment (codeinthehole) [inline tests/test_ranges.py:295]: I gave up trying to get this one to pass. Error is:
```
tests/test_ranges.py:295: error: Unsupported left operand type for | ("None")  [operator]
```

File: xocto/storage/storage.py [hunk 2]
```diff
@@ -43,8 +43,6 @@
     from mypy_boto3_s3.client import S3Client
 
 
-StorageFile = StreamingBody
```
Comment (codeinthehole) [inline xocto/storage/storage.py:46]: Mypy didn't like this - I needed to change the return type to `StreamingBody`
```
xocto/storage/storage.py:294: error: Variable "xocto.storage.storage.StorageFile" is not valid as a type  [valid-type]
xocto/storage/storage.py:294: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
xocto/storage/storage.py:539: error: Variable "xocto.storage.storage.StorageFile" is not valid as a type  [valid-type]
xocto/storage/storage.py:539: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
xocto/storage/storage.py:544: error: StorageFile? has no attribute "read"  [attr-defined]
xocto/storage/storage.py:1080: error: Variable "xocto.storage.storage.StorageFile" is not valid as a type  [valid-type]
xocto/storage/storage.py:1080: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
xocto/storage/storage.py:1101: error: StorageFile? has no attribute "read"  [attr-defined]
xocto/storage/storage.py:1349: error: Variable "xocto.storage.storage.StorageFile" is not valid as a type  [valid-type]
xocto/storage/storage.py:1349: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
```

**Review by codeinthehole — COMMENTED** [review]
Notes

**Review by adsultana — APPROVED** [review]

File: tests/test_ranges.py (hunk 1 above)
Comment (codeinthehole) [reply tests/test_ranges.py:295]: Good call - I'll use a comment.

File: xocto/types.py [hunk 3]
```diff
@@ -37,19 +37,19 @@ class Comparable(Protocol[T]):
...
+    def __le__(self, other: Any) -> bool:
         ...
 
-    def __gt__(self, other) -> bool:
+    def __gt__(self, other: Any) -> bool:
         ...
 
-    def __ge__(self, other) -> bool:
+    def __ge__(self, other: Any) -> bool:
         ...
```
[the remark this answers is not in the document]
Comment (codeinthehole) [reply xocto/types.py:53]: I tried that but get:
```
xocto/types.py:34: error: Covariant type variable "T" used in protocol where contravariant one is expected  [misc]
xocto/types.py:40: error: Argument 1 of "__eq__" is incompatible with supertype "object"; supertype defines the argument type as "object"  [override]
xocto/types.py:40: note: This violates the Liskov substitution principle
xocto/types.py:40: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
xocto/types.py:40: note: It is recommended for "__eq__" to work with arbitrary objects, for example:
xocto/types.py:40: note:     def __eq__(self, other: object) -> bool:
xocto/types.py:40: note:         if not isinstance(other, Comparable):
xocto/types.py:40: note:             return NotImplemented
xocto/types.py:40: note:         return <logic to compare two Comparable instances>
xocto/types.py:40: error: Cannot use a covariant type variable as a parameter  [misc]
xocto/types.py:43: error: Cannot use a covariant type variable as a parameter  [misc]
xocto/types.py:46: error: Cannot use a covariant type variable as a parameter  [misc]
xocto/types.py:49: error: Cannot use a covariant type variable as a parameter  [misc]
xocto/types.py:52: error: Cannot use a covariant type variable as a parameter  [misc]
xocto/numbers.py:114: error: Unsupported operand types for > ("Comparable[T]" and "Comparable[T]")  [operator]
xocto/numbers.py:115: error: Unsupported operand types for < ("Comparable[T]" and "Comparable[T]")  [operator]
xocto/numbers.py:117: error: Unsupported operand types for > ("Comparable[T]" and "Comparable[T]")  [operator]
Found 10 errors in 2 files (checked 35 source files)
```
and by that point, I had run out of energy for this. You're welcome to pick it up if you can find the right way of placating Mypy.

[the remark this answers is not in the document]
Comment (adsultana) [reply xocto/types.py:53]: Yeah, I had a feeling that would be the case. Co/contravariance is one of those concepts I've never taken the time to fully grok, so will store this under the "hopefully someday" bucket of things to do.

**Pull request merged**